src/lib/provable/gadgets/gadgets.ts

```diff
@@ -867,6 +892,43 @@ const Gadgets = {
    *    */
   addMod32,
 
+  /**
+   * Division modulo 2^64. The operation decomposes a {@link Field} element in the range [0, 2^128) into two 64-bit limbs, `remainder` and `quotient`, using the following equation: `n = quotient * 2^64 + remainder`.
+   *
+   * **Note:** The gadget acts as a proof that the input is in the range [0, 2^128). If the input exceeds 128 bits, the gadget fails.
+   *
+   * Asserts that both `remainder` and `quotient` are in the range [0, 2^64) using {@link Gadgets.rangeCheck64}.
+   *
+   * @example
+   * ```ts
+   * let n = Field((1n << 64n) + 8n)
+   * let { remainder, quotient } = Gadgets.divMod64(n);
+   * // remainder = 8, quotient = 1
+   *
+   * n.assertEquals(quotient.mul(1n << 64n).add(remainder));
+   * ```
+   */
+  divMod64,
+
+  /**
+   * Addition modulo 2^64. The operation adds two {@link Field} elements in the range [0, 2^128] and returns the result modulo 2^64.
```

Comment:
TBH, I am not sure which one is right. The ranges in `addMod32` comment block doesn't match too. I took the comment for `addMod64` from `addMod32` since it's the exact same function but modulo 2^64.

Reply:
first of all, all ranges are always of the form `[a, b)` i.e. open on the upper end

second, 128 bits is wrong. it uses a 1-bit range check on the quotient. for that to work, the addition result has to be at most 65 bits

also, you say further below that "the gadget assumes both inputs to be in the range [0, 2^64)". that will definitely work an is the use case, so stick with that here in the description as well

Reply:
fixed!

Reply:
the inputs are in the range [0, 2^64) though, not [0, 2^65)

Reply:
right, it should be correct now.